### PR TITLE
LLM prompts synthetic jobs

### PR DESCRIPTION
**Pull request by emnoor-reef**
The nice abstraction in job generator classes and `batch_run.py` was broken in this PR to not slow down the process too much.

- special cased in `batch_run._generate_jobs` for the llama jobs to fetch all the prompts related db objects in one go (instead of fetching them one by one in the job generator `ainit` method)
- special cased in `batch_run._score_jobs` for the llama jobs to fetch the results from s3 for all the completed jobs (instead of fetching them one by one in the job generator `verify` method)

### CODE REVIEW

File: validator/app/src/compute_horde_validator/validator/synthetic_jobs/generator/llama_prompts.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+)
+
+
+@dataclass(frozen=True, slots=True)
+class PromptAnswer:
+    prompt: str
+    answer: str
+
+
+class LlamaPromptsSyntheticJobGenerator(BaseSyntheticJobGenerator):
```
Comment (andreea-popescu-reef) [inline validator/app/src/compute_horde_validator/validator/synthetic_jobs/generator/llama_prompts.py:26]: maybe drop llama from the name as we are not necessarily using that
`WorkloadSolverJobGenerator` or `PromptSampleJobGenerator` might work better? as long as it doesn't clash with the https://github.com/backend-developers-ltd/ComputeHorde/pull/235/files#diff-d1fdc32a1526206707fb838e326e61cbb10d02cb0fb8aee0fd4ccfe5addb6412R6 naming
maybe also rename the filename?

File: validator/app/src/compute_horde_validator/validator/synthetic_jobs/generator/llama_prompts.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+    def docker_image_name(self) -> str:
+        return "TODO"
+
+    def docker_run_options_preset(self) -> str:
+        return "nvidia_all"
+
+    def docker_run_cmd(self) -> list[str]:
+        return [
+            "--temperature=0.5",
```
Comment (olzhasar-reef) [inline validator/app/src/compute_horde_validator/validator/synthetic_jobs/generator/llama_prompts.py:75]: I think we should discuss if we possibly want to have these in the dynamic config

File: validator/app/src/compute_horde_validator/validator/synthetic_jobs/generator/llama_prompts.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+            response.raise_for_status()
+            self.prompt_answers = [
+                PromptAnswer(prompt, answer) for prompt, answer in response.json().items()
+            ]
+
+    def verify(self, msg: V0JobFinishedRequest, time_took: float) -> tuple[bool, str, float]:
+        if self.prompt_answers is None:
+            raise RuntimeError("_download_answers must be called before calling verify")
+
+        for prompt in self.prompts:
```
Comment (olzhasar-reef) [inline validator/app/src/compute_horde_validator/validator/synthetic_jobs/generator/llama_prompts.py:115]: We should consider avoiding a nested loop here by using a mapping :)

File: validator/app/src/compute_horde_validator/validator/synthetic_jobs/generator/llama_prompts.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+            getattr(prompt_sample, "_prefetched_objects_cache", {}).get("prompts") is not None
+        ), _PROMPT_SAMPLE_RELATED_NOT_CACHED
+
+        self.prompt_sample: PromptSample = prompt_sample
+        self.prompts: list[Prompt] = list(self.prompt_sample.prompts.all())
+
+        self.input_filename = str(uuid.uuid4()) + ".txt"
+        self.s3_output_key = str(uuid.uuid4()) + ".json"
+        self.s3_output_prefix = "solved/"
+        self.s3_output_bucket = settings.S3_BUCKET_NAME_ANSWERS
```
Comment (olzhasar-reef) [inline validator/app/src/compute_horde_validator/validator/synthetic_jobs/generator/llama_prompts.py:42]: This bucket is used to store answers that we got on our trusted hardware. Using the same one for synthetic jobs later is maybe okay, but feels terrifying for some reason :)

File: compute_horde/compute_horde/executor_class.py [hunk 5]
```diff
@@ -6,6 +6,7 @@
 class ExecutorClass(StrEnum):
     spin_up_4min__gpu_24gb = "spin_up-4min.gpu-24gb"
     always_on__gpu_24gb = "always_on.gpu-24gb"
+    always_on__llama = "always_on.llama"
```
[the remark this answers is not in the document]
Comment (mzukowski-reef) [reply compute_horde/compute_horde/executor_class.py:9]: this is good point - results are different on different hardware, so we want to say: a6000

File: validator/app/src/compute_horde_validator/validator/synthetic_jobs/generator/llm_prompts.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    def _url_for_download(self) -> str:
+        return get_public_url(
+            key=self.s3_output_key,
+            bucket_name=self.s3_output_bucket,
+            prefix=self.s3_output_prefix,
+        )
+
+    def timeout_seconds(self) -> int:
+        # TODO: ???
+        return 80
```
Comment (emnoor-reef) [inline validator/app/src/compute_horde_validator/validator/synthetic_jobs/generator/llm_prompts.py:54]: what should be this value?

**Comment by emnoor-reef** [comment]
Merged PR https://github.com/backend-developers-ltd/ComputeHorde/pull/240 here for simpler testing and fixes. There were some pending discussions on that PR that still need to be addressed.

File: validator/app/src/compute_horde_validator/validator/synthetic_jobs/batch_run.py [hunk 7]
```diff
@@ -761,16 +764,61 @@ async def _close_client(ctx: BatchContext, miner_hotkey: str) -> None:
...
         for executor_class, count in executors.items():
             job_generators = []
             for _ in range(count):
-                job_generator = await current.synthetic_job_generator_factory.create(executor_class)
+                kwargs = {}
+                if executor_class == ExecutorClass.always_on__llm__a6000:
+                    if prompt_samples_iter is None:
+                        logger.warning("No llm prompt samples available, skipping llm job")
+                        continue
+                    prompt_sample = next(prompt_samples_iter)
```
Comment (mzukowski-reef) [inline validator/app/src/compute_horde_validator/validator/synthetic_jobs/batch_run.py:811]: calling `next` without default value is.... dangerous :)

File: validator/app/src/compute_horde_validator/validator/synthetic_jobs/generator/llm_prompts.py (hunk 6 above)
Comment (mzukowski-reef) [reply validator/app/src/compute_horde_validator/validator/synthetic_jobs/generator/llm_prompts.py:53]: 41
42
42
41
42
42
42
42

so 48?

**Pull request merged**